magenta-lib/src/main/scala/magenta/tasks/AWS.scala

```diff
@@ -81,6 +82,24 @@ object S3 extends AWS {
     }
     bucketName
   }
+
+  def resolveBucket(bucket: Bucket, ssmClient: => SsmClient, reporter: => DeployReporter): String = {
+    bucket match {
+      case BucketByName(name) => name
+      case bucketBySsm: BucketBySsmKey =>
+        val resolvedBucket = bucketBySsm.resolve(ssmClient)
```

Comment:
just a thought to reduce the number of methods calling each other and make things more cohesive:
In general you can implement polymorphism by either pattern matching or inheritance (or typeclasses if you have a special use case) wheras here we pattern match and then call a method.
```suggestion
        val resolvedBucket = SSM.getParameter(ssmClient, bucketBySsm.ssmKey)
```
then you can remove the whole resolve method from the BucketBySsmKey thereby making it into an ADT

Reply:
Some of the complexity comes from some things happening when the task list is being made and some during task execution but you're quite right - this is confusing in the way it mixes two approaches incoherently. I'll look to simplify this.